### ISSUE TITLE
Practical 4A: Backpropagation Algorithm

### ISSUE THREAD

**Issue by karinkato2103** [karinkato2103/soft#7]
import numpy as np

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

class BackpropNN:
    def __init__(self):
        rng = np.random.RandomState(2)
        self.w1 = rng.normal(scale=0.5, size=(2, 2))
        self.b1 = np.zeros(2)
        self.w2 = rng.normal(scale=0.5, size=(2, 1))
        self.b2 = np.zeros(1)
        self.lr = 0.5

    def train(self, X, y, epochs=20000):
        for _ in range(epochs):
            h = sigmoid(X.dot(self.w1) + self.b1)
            o = sigmoid(h.dot(self.w2) + self.b2)

            error = o - y

            d2 = error * o * (1 - o)
            d1 = d2.dot(self.w2.T) * h * (1 - h)

            self.w2 -= self.lr * h.T.dot(d2)
            self.b2 -= self.lr * d2.sum(axis=0)

            self.w1 -= self.lr * X.T.dot(d1)
            self.b1 -= self.lr * d1.sum(axis=0)

    def predict(self, X):
        h = sigmoid(X.dot(self.w1) + self.b1)
        o = sigmoid(h.dot(self.w2) + self.b2)
        return (o >= 0.5).astype(int)

if __name__ == "__main__":
    X = np.array([[0,0],[0,1],[1,0],[1,1]])
    y = np.array([[0],[1],[1],[0]])

    nn = BackpropNN()
    nn.train(X, y)

    print(nn.predict(X).reshape(-1))
